### ISSUE TITLE
Provide the ability to remove code needed only in dev mode from the production bundle

### ISSUE THREAD

**Issue by markostanimirovic** [angular/angular#51175]
### Which @angular/* package(s) are relevant/related to the feature request?

core

### Description

Several NgRx libraries have features/checks that are only used in development mode (e.g. store runtime checks, component store lifecycle hooks check, store devtools, etc).

### Proposed solution

Expose the `ngDevMode`/`isNgDevMode` variable to the public API, so it can be used in Angular libraries that have dev mode-related functionalities, but also in user codebases if needed.

This feature could significantly reduce the size of production bundles in many Angular codebases.

### Alternatives considered

Currently, we can use `ngDevMode` to remove dev mode-related logic from the production bundle in the following way:

```ts
declare const ngDevMode: boolean;
const isNgDevMode = typeof ngDevMode === 'undefined' || !!ngDevMode;

// ...

if (isNgDevMode) {
  console.warn('@ngrx/store: ...');
}
```

However, this global variable is not part of the public API and it can be risky to rely on it.

**Comment by JeanMeche** [comment]
Hi Marko, we already have [`isDevMode`](https://angular.io/api/core/isDevMode). have you tried it ?

**Comment by markostanimirovic** [comment]
Yes, we're using the `isDevMode` function, but we would like to eliminate "dead" code from the production bundle which is possible with `ngDevMode`.

**Comment by arturovt** [comment]
Terser is not able to tree-shake the code if the function, returning a statically analyzeble value, is used in more than 1 module. Terser is not analyzing the entire project's dependency graph, that's why `ngDevMode` is defined everywhere in the Angular code, which means it doesn't perform cross-module tree-shaking. You can check the following example:

```js
// file-1.js
import { isDevMode } from './shared';

isDevMode() && console.log('file-1');

// file-2.js
import { isDevMode } from './shared';

isDevMode() && console.log('file-2');

// shared.js
export function isDevMode() {
  return !!ngDevMode;
}
```
```js
// webpack.config.js
const TerserPlugin = require('terser-webpack-plugin');

module.exports = {
  entry: ['./src/file-1.js', './src/file-2.js'],
  optimization: {
    minimize: true,
    minimizer: [
      new TerserPlugin({
        terserOptions: {
          compress: {
            global_defs: {
              ngDevMode: false,
            },
          },
        },
      }),
    ],
  },
};
```

The output code:

```js
(() => {
  'use strict';
  var e = {
      768: (e, r, o) => {
        function t() {
          return !1;
        }
        o.d(r, { X: () => t });
      },
    },
    r = {};
  function o(t) {
    var n = r[t];
    if (void 0 !== n) return n.exports;
    var i = (r[t] = { exports: {} });
    return e[t](i, i.exports, o), i.exports;
  }
  (o.d = (e, r) => {
    for (var t in r)
      o.o(r, t) &&
        !o.o(e, t) &&
        Object.defineProperty(e, t, { enumerable: !0, get: r[t] });
  }),
    (o.o = (e, r) => Object.prototype.hasOwnProperty.call(e, r)),
    (0, o(768).X)() && console.log('file-1'),
    (0, o(768).X)() && console.log('file-2');
})();
```

**Comment by JeanMeche** [comment]
> 1. export the `ngDevMode` as a variable,

`ngDevMode` is not a conventional constant. It's set the by the cli and declared as global inside the framework. 

https://github.com/angular/angular/blob/3a59de681fd9899a40c49b3bce6f101d3d1b95cd/packages/core/src/util/ng_dev_mode.ts#L11-L26

**Comment by maxime1992** [comment]
I've been using a little trick in my case with ngrx which is to have 2 files in which I define what I want, for example the redux dev tool import. In dev mode I import it, while in the other file I don't. I then use the angular cli swapping file capability to use the dev one at serve time and the prod one (with nothing it in except an empty array) at build time.

And because I feel like I'm absolutely not clear (end of day sorry :upside_down_face:) here's what I've got:

`ngrx.ts`:

```typescript
import { ModuleWithProviders } from '@angular/core';
import { StoreDevtoolsModule } from '@ngrx/store-devtools';

// this file will be replaced by the prod version at build time!
// this is to make sure that we don't include `store-devtools` in our prod bundle
// https://ngrx.io/guide/store-devtools/recipes/exclude

export const ngrxModules: ModuleWithProviders<any>[] = [StoreDevtoolsModule.instrument({ maxAge: 25 })];
```

`ngrx.prod.ts`:

```typescript
import { ModuleWithProviders } from '@angular/core';

export const ngrxModules: ModuleWithProviders<any>[] = [];
```

And in my Angular CLI config file: 

```
          "fileReplacements": [
            {
              "replace": "apps/my-app/src/app/build-specifics/ngrx.ts",
              "with": "apps/my-app/src/app/build-specifics/ngrx.prod.ts"
            }
          ],
```

Now, while there's a work around for this... I have to admit that this feature would be a really nice to have. It's quite convoluted and error prone to swap files at build time. Meanwhile, in case anyone needs a workaround to make sure you don't bundle the entire code for the redux dev tool or others, you know where to look :)

**Comment by simeyla** [comment]
I'm using the `environment.ts` [file](https://angular.io/guide/build) (which has sort of gone out of favor, but exists today) to achieve this.

There is a missed opportunity in the current docs though. If you choose to enable `environment.ts`, the initial config defines `production: boolean` which can be used to conditionally run (or not run) code.

```
export const environment = {
  production: true
};
```

Unfortunately (and this should probably be mentioned in the docs ) any code using `environment.production` can't be tree shaken away because while `environment` may be a constant, the javascript object itself is mutable and there's nothing stopping you from setting `environment.production = false` at runtime. Even if you add `as const` that makes no difference since only Typescript cares about that.

I chose to solve this by defining two top level constants in `environment.ts` (reversed in `environment.prod.ts`). Code inside `if (DEV_MODE)` is now able to be tree shaken in production. I personally prefer the explicitness of two values.

```
export const DEV_MODE = true;
export const IS_PRODUCTION = false;
```

I wouldn't ever want to use `ngDevMode` in application code - but for libraries it makes more sense.

**Comment by JeanMeche** [comment]
The usecase for`ngDevMode` is definitely oriented toward libraries that want to provide a similar behaviour as Angular : 

Explicit logs and errors in devMode & only critical errors in prodMode (and less verbose).

**Comment by yharaskrik** [comment]
> @simeyla  The usecase for`ngDevMode` is definitely oriented toward libraries that want to provide a similar behaviour as Angular : 
> 
> Explicit logs and errors in devMode & only critical errors in prodMode (and less verbose).  

This. Please. I have been wanting to do this for a while! Will help us library authors provide really clear and complete verbose logging but then have prod bundles be as small as possible.

**Comment by alan-agius4** [comment]
**Note**: `ngDevMode` is considered as a private API and is not intended but to be used by 3rd parties.

**Comment by JeanMeche** [comment]
Yes, that's stated by @markostanimirovic in his post. The request here would be to make it public so library maintainers can officially rely on it.

**Comment by JeanMeche** [comment]
Shouldn't we close this issue in favor of the one on the CLI repo ? angular/angular-cli#23738

**Comment by simeyla** [comment]
@JeanMeche If so it needs a new title. The title of *this* question is much more suitable and easy to find.

I did want to add an alternative approach (not sure whether or not this works for library authors since I've never created such a library).

In `angular.json` you can now add constants under 'define'. This is similar to the [webpack DefinePlugin] and [esbuild equivalent](https://esbuild.github.io/api/#define) (https://webpack.js.org/plugins/define-plugin/). However since it's now in angular.json it will work with both build systems.

I couldn't immediately find official mention of `options.define` in the angular.json docs, but here's a article https://netbasal.com/explore-angular-clis-define-option-effortless-global-identifier-replacement-f08fec7d9243

**Comment by alan-agius4** [comment]
@JeanMeche, actually the CLI issue is rather working as expected, the missing part is having a public API similar to `ngDevMode` that is treeshakable.

Worth to mention that the story around `ngDevMode` also changed over the years as originally  this was not meant to be used to as a token to remove code not even in the framework code case, but rather more for dev mode runtime stats.

`options.define` can definitely be used as a valid alternative. @markostanimirovic, mind opening a new issue in this repo for a request to document the `define` option. Likely it should be included in https://angular.dev/reference/configs/workspace-config#complex-configuration-value

**Comment by k3nsei** [comment]
@alan-agius4 @simeyla But how this `options.define` would work for library authors? As when we are building libraries we want to include this code when posting for example to NPM. We only want to remove this parts of code when application that using library is building. Angular is opinionated and this token also should be common for all community. To not introduce 3, 5 or 10 different tokens with `options.define` as every library author would choose different name.

**Comment by jkrems** [comment]
Would it be reasonable to support `import.meta.env.DEV`? It looks like that has some buy-in from a couple of bundlers (rspack, vite, bun to some extend).

**Comment by JeanMeche** [comment]
#28531 is what Jan is takling about :)

**Comment by Platonn** [comment]
As a library author, we'd like to provide verbose helpful error messages in our logic in case of errors/misconfiguration/etc from our library users. Unfortunately the end users of our library in their app's production build pay the price of our verbose logs which are meant for improving their DX only in the dev mode. Bundling those KB of string error messages in production is unjustified and which contributes a bit to degrading Core Web Vitals for our library consumers.

Example:

```ts
class Service1 {
  /* ... */
  validateConfig(){
    if(isDevMode) { // we hope to tree-shake all of it in prod build
      if (/* validation rules */) {
        this.logger(/*
          very..............................................
          long..............................................
          verbose..............................................
          message..............................................
          with..............................................
          details..............................................
          what went wrong..............................................
          and advices how to fix it.
      /*);
    }
  }
}
```

We have many places in our library's code with such logs. Those logs are helpful for humans, but also for AI agents, which speedup work for our library consumes.

If possible, we'd really love to have a way to eaisly mark branches of our code as "devmode only" and have those branches excluded from the production build of the apps of our library consumers, even if it would require architecting our code differently or applying strange workarounds. Of course, ideally would be to use something as convenient as `if (ngDevMode)`, if possible.

**Comment by pburgmer** [comment]
> As a library author, we'd like to provide verbose helpful error messages in our logic in case of errors/misconfiguration/etc from our library users. Unfortunately the end users of our library in their app's production build pay the price of our verbose logs which are meant for improving their DX only in the dev mode. Bundling those KB of string error messages in production is unjustified and which contributes a bit to degrading Core Web Vitals for our library consumers.
> 
> Example:
> 
> class Service1 {
>   /* ... */
>   validateConfig(){
>     if(isDevMode) { // we hope to tree-shake all of it in prod build
>       if (/* validation rules */) {
>         this.logger(/*
>           very..............................................
>           long..............................................
>           verbose..............................................
>           message..............................................
>           with..............................................
>           details..............................................
>           what went wrong..............................................
>           and advices how to fix it.
>       /*);
>     }
>   }
> }
> We have many places in our library's code with such logs. Those logs are helpful for humans, but also for AI agents, which speedup work for our library consumes.
> 
> If possible, we'd really love to have a way to eaisly mark branches of our code as "devmode only" and have those branches excluded from the production build of the apps of our library consumers, even if it would require architecting our code differently or applying strange workarounds. Of course, ideally would be to use something as convenient as `if (ngDevMode)`, if possible.

I have the same situation
